framework/Web/THttpRequest.php

```diff
@@ -767,43 +768,73 @@ protected function parseUrl()
+
+		// Prepend parameters extracted from the URL so they are first in the parameters array
+		$this->_items = array_merge($urlParams, $_GET, $_POST);
+		$urlParams = $this->onResolveRequest($serviceIDs, $original = $urlParams);
+		if (is_string($urlParams)) {
+			$serviceID = $urlParams;
+		} elseif ($urlParams !== false) {
+			if ($urlParams !== $original) {
+				$this->_items = array_merge($urlParams, $_GET, $_POST);
 			}
-			$this->_items = array_merge($_GET, $_POST);
-			$this->_requestResolved = true;
-			foreach ($serviceIDs as $serviceID) {
-				if ($this->contains($serviceID)) {
-					$this->setServiceID($serviceID);
-					$this->setServiceParameter($this->itemAt($serviceID));
-					return $serviceID;
+			if ($this->_resolveMethod == THttpRequestResolveMethod::ServiceOrder) {
+				foreach ($serviceIDs as $id) {
+					if ($this->contains($id)) {
+						$serviceID = $id;
+						break;
+					}
 				}
-			}
-		} else {
-			// THttpRequestResolveMethod::ParameterOrder
-			// Prepend parameters extracted from the URL so they gets first in the parameters array
-			$this->_items = array_merge($getParams, $_GET, $_POST);
-			$this->_requestResolved = true;
-			foreach ($this->_items as $serviceID => $value) {
-				if (in_array($serviceID, $serviceIDs)) {
-					$this->setServiceID($serviceID);
-					$this->setServiceParameter($this->itemAt($serviceID));
-					return $serviceID;
+			} else {
+				// THttpRequestResolveMethod::ParameterOrder
+				foreach ($this->_items as $key => $value) {
+					if (in_array($key, $serviceIDs)) {
+						$serviceID = $key;
+						break;
+					}
 				}
 			}
 		}
-		return null;
+		$this->_requestResolved = true;
+		if ($serviceID !== null) {
+			$this->setServiceID($serviceID);
+			$this->setServiceParameter($this->itemAt($serviceID));
+		}
+		return $serviceID;
+	}
+
+	/**
+	 * Raised when resolving the request.  This is for application specific request
+	 * resolution.  Event Handlers can return `false` for no service, a `string` of the
+	 * service id to use, or an array replacing the $urlParams.  By returning null,
+	 * the result is skipped and has no effect.  Event Handler results are stacked and
+	 * results are popped until a valid result is found.  The only invalid result is
+	 * if a string is provided that is not in the $serviceIDs.
+	 * The lower priorities of event handlers are more important and so should be last.
+	 * @param array $serviceIDs The Service IDs of the application.
+	 * @param array $urlParams The decoded url parameters to be updated or replaced
+	 * @return array|false|string
+	 */
+	public function onResolveRequest(array $serviceIDs, array $urlParams): false|string|array
+	{
+		$results = $this->raiseEvent('onResolveRequest', $this, new THttpRequestParameter($serviceIDs, $urlParams), TEventResults::EVENT_REVERSE);
+		while (count($results)) {
+			$return = array_pop($results);
+			if (is_string($return) && in_array($return, $serviceIDs) || is_array($return) || $return === false) {
```

Comment:
Please add some parenthesis to avoid confusion in operator precedence.
Where is the "null" case handled? (By returning null, the result is skipped and has no effect) I guess it just loops over all non-matched data types, including null.

Reply:
The results could be filtered for null by including the `TEventResults::EVENT_REVERSE  | TEventResults::EVENT_RESULT_FILTER` flag.  You are correct, null is simply skipped, whether filtered or just ignored.  I like adding the filter.

Also, && is always logically combined first before ||.  The operator order is specified in PHP.  with this knowledge, it's not confusing.  But if you think so, OK.

Reply:
my bad.  that's right, the event filter cannot be used because it will filter out the `false` values as well.  It is the case of simply skipping over data that doesn't match.

Only strings in the serviceIDs, arrays to update the urlParam and false.  it does run down the array of results for the first valid result

Reply:
I'm updating the parenthesis but I don't think it's needed.

Reply:
> I'm updating the parenthesis but I don't think it's needed.

You are correct, it's just my paranoia due to php changing the way operators worked in the past (eg. ternary operators in 7.4)

Reply:
I know what you mean.  some things are wildly more nuanced than they appear on the surface.  ternary and order of operation can be like that.

BTW, I have this double ternary operator that is very interesting to see, at least to me when I first put it together.
eg `($endian === null) ? $endian ? 'N' : 'n' : 'l'`